File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/statement/Statement.java
package org.seasar.doma.jdbc.criteria.statement;

import java.util.function.Consumer;
import org.seasar.doma.jdbc.Sql;

/**
 * Represents the SQL statement built by the Criteria API.
 *
 * @param <RESULT> the result of the execution
 */
public interface Statement<RESULT> extends Buildable<Statement<RESULT>> {

  /**
   * Executes the SQL statement.
   *
   * @return the result
   */
  RESULT execute();

  @Override
  default Statement<RESULT> peek(Consumer<Sql<?>> consumer) {
    return Buildable.super.peek(consumer);
  }
}

File: doma-criteria/src/main/kotlin/org/seasar/doma/criteria/Statement.kt
package org.seasar.doma.criteria

import org.seasar.doma.jdbc.Config
import org.seasar.doma.jdbc.PreparedSql
import org.seasar.doma.jdbc.Sql
import org.seasar.doma.jdbc.command.Command

interface Statement<RESULT> {
    fun execute(config: Config): RESULT
    fun asSql(config: Config): Sql<*>
}

abstract class AbstractStatement<RESULT> : Statement<RESULT> {
    override fun execute(config: Config): RESULT {
        val (command, _) = commandAndSql(config)
        return command.execute()
    }

    override fun asSql(config: Config): Sql<*> {
        val (_, sql) = commandAndSql(config)
        return sql
    }

    protected abstract fun commandAndSql(config: Config): Pair<Command<RESULT>, PreparedSql>
}
